numpy/core/arrayprint.py

```diff
@@ -816,35 +824,48 @@ def recurser(index, hanging_indent, curr_width):
             s, line = _extendLine_pretty(
                 s, line, word, elem_width, hanging_indent, legacy)
 
-            s += line
+            s.append(''.join(line))
 
         # other axes - insert newlines between rows
         else:
-            s = ''
+            s = []
             line_sep = separator.rstrip() + '\n'*(axes_left - 1)
 
             for i in range(leading_items):
-                nested = recurser(index + (i,), next_hanging_indent, next_width)
-                s += hanging_indent + nested + line_sep
+                nested = recurser(index + (i,), next_hanging_indent,
```

Comment:
IMHO, `recurser` returns a string which gets added to `s` in the previous call. In this process of addition, first `nested` is copied and appended to `s`. Then this `s` is returned to previous call where same thing happens. This leads to expensive copies in the earlier calls leading to heavy memory requirements and slow downs.
For example, in the below output, each line represents the length of `word`/`nested` obtained from call to `recurser`. In the earlier calls (i.e., in the below lines) the length explodes.
<details>
<summary>Increasing String lengths from calls to `recurser`</summary>
<pre>
1.22.0.dev0+560.gd61fdc7d4
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
2868
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
2868
5748
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
2868
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
2868
5748
11508
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
2868
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
2868
5748
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
2868
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
15
15
33
15
15
33
78
15
15
33
15
15
33
78
168
348
708
1428
2868
5748
11508
|   size |   Time (s) |   Memory (bytes) |
|--------|------------|------------------|
|     10 |      0.009 |         31543296 |
</pre>
</details>